File: ClipCascade_Mobile/src/fragmentUtils.js
'use strict';

// Shared, dependency-free helpers for P2P clipboard fragmentation.
// Used by StartForegroundService.js to:
// - split UTF-8 text into fragments without corrupting multi-byte sequences
//   (the previous naive byte slicing turned characters straddling a fragment
//   boundary into U+FFFD replacement characters), and
// - validate remote fragment metadata before allocating memory for
//   reassembly (a peer could previously request an arbitrary number of
//   pre-allocated fragments).

// Absolute ceiling on fragments accepted for a single stream, regardless of
// any configured local size limit. At the default FRAGMENT_SIZE of 15 KiB
// this permits payloads up to ~1 GiB while keeping the pre-allocation for
// reassembly bounded (an unclamped peer-supplied totalFragments could
// previously trigger a multi-GB allocation).
const MAX_TOTAL_FRAGMENTS_HARD = 65536;

// Lower bound so that even a small configured limit still accepts ordinary
// payloads; the size-based rejection itself is handled separately by the
// combinedRawPayloadSizeInBytes check.
const MAX_TOTAL_FRAGMENTS_FLOOR = 4096;

// Safety margin over ceil(limit / fragment_size) for the derived cap.
const MAX_TOTAL_FRAGMENTS_MARGIN = 16;

const UTF8_CONTINUATION_MASK = 0xc0;
const UTF8_CONTINUATION_PATTERN = 0x80;

/**
 * Split `text` into fragments of at most `size` UTF-8 bytes each, where
 * every fragment is a valid, standalone UTF-8 string and the fragments
 * concatenated in order reproduce `text` exactly.
 *
 * @param {string} text the string to fragment
 * @param {number} size maximum byte size of each fragment (>= 4)
 * @returns {string[]} the fragments (empty array for empty input)
 * @throws {Error} if size is too small to guarantee progress
 */
function utf8SafeChunks(text, size) {
  if (!Number.isInteger(size) || size < 4) {
    // A UTF-8 codepoint is at most 4 bytes, so size >= 4 guarantees the
    // boundary back-off below always makes progress.
    throw new Error('fragment size must be an integer >= 4');
  }

  const data = new TextEncoder().encode(text);
  const total = data.length;
  const chunks = [];
  const decoder = new TextDecoder();
  let start = 0;

  while (start < total) {
    let end = Math.min(start + size, total);
    if (end < total) {
      // Back off over continuation bytes so the cut lands between codepoints.
      while (end > start && (data[end] & UTF8_CONTINUATION_MASK) === UTF8_CONTINUATION_PATTERN) {
        end -= 1;
      }
    }
    // Aligned chunks are valid UTF-8 by construction; TextDecoder throws
    // (non-fatal mode is not set) if that invariant is ever violated.
    chunks.push(decoder.decode(data.subarray(start, end)));
    start = end;
  }

  return chunks;
}

/**
 * Upper bound on the number of fragments a single inbound stream may declare
 * before it is rejected. When a positive local clipboard size limit is
 * configured, the bound is derived from it (so legitimate transfers under the
 * limit always fit); otherwise the hard ceiling applies.
 *
 * @param {number} fragmentSize maximum byte size of one fragment
 * @param {number|null} localLimitBytes configured local limit in bytes (or a
 *   non-positive/undefined value meaning "unlimited")
 * @returns {number}
 */
function maxTotalFragments(fragmentSize, localLimitBytes) {
  if (Number.isFinite(localLimitBytes) && localLimitBytes > 0) {
    const derived =
      Math.ceil(localLimitBytes / fragmentSize) + MAX_TOTAL_FRAGMENTS_MARGIN;
    return Math.max(
      MAX_TOTAL_FRAGMENTS_FLOOR,
      Math.min(derived, MAX_TOTAL_FRAGMENTS_HARD),
    );
  }
  return MAX_TOTAL_FRAGMENTS_HARD;
}

/**
 * True when totalFragments / fragmentIndex are well-formed integers
 * describing a position inside the stream. Rejects floats, NaN, and
 * out-of-range indices.
 *
 * @param {*} totalFragments
 * @param {*} fragmentIndex
 * @returns {boolean}
 */
function isValidFragmentMetadata(totalFragments, fragmentIndex) {
  if (!Number.isInteger(totalFragments) || !Number.isInteger(fragmentIndex)) {
    return false;
  }
  if (totalFragments < 1) {
    return false;
  }
  return fragmentIndex >= 0 && fragmentIndex < totalFragments;
}

module.exports = {
  MAX_TOTAL_FRAGMENTS_HARD,
  MAX_TOTAL_FRAGMENTS_FLOOR,
  MAX_TOTAL_FRAGMENTS_MARGIN,
  utf8SafeChunks,
  maxTotalFragments,
  isValidFragmentMetadata,
};

File: ClipCascade_Desktop/src/core/fragment_utils.py
"""
Shared, dependency-free helpers for P2P clipboard fragmentation.

Used by p2p_manager to:
- split UTF-8 text into fragments without corrupting multi-byte sequences
  (the previous naive byte slicing silently dropped characters that
  straddled a fragment boundary), and
- validate remote fragment metadata before allocating memory for
  reassembly (a peer could previously request an arbitrary number of
  pre-allocated fragments).
"""

import math

# Absolute ceiling on fragments accepted for a single stream, regardless of
# any configured local size limit. At the default FRAGMENT_SIZE of 15 KiB
# this permits payloads up to ~1 GiB while keeping the pre-allocation for
# reassembly bounded (an unclamped peer-supplied totalFragments could
# previously trigger a multi-GB allocation).
MAX_TOTAL_FRAGMENTS_HARD = 65536

# Lower bound so that even a small configured limit still accepts ordinary
# payloads; the size-based rejection itself is handled separately by the
# combinedRawPayloadSizeInBytes check.
MAX_TOTAL_FRAGMENTS_FLOOR = 4096

# Safety margin over ceil(limit / fragment_size) for the derived cap.
MAX_TOTAL_FRAGMENTS_MARGIN = 16

# UTF-8 continuation bytes are of the form 0b10xxxxxx.
_UTF8_CONTINUATION_MASK = 0xC0
_UTF8_CONTINUATION_PATTERN = 0x80


def utf8_safe_chunks(text: str, size: int) -> list:
    """
    Split `text` into fragments of at most `size` UTF-8 bytes each, where
    every fragment is a valid, standalone UTF-8 string and the fragments
    concatenated in order reproduce `text` exactly.

    Args:
        text: The string to fragment.
        size: Maximum byte size of each fragment (must be >= 4 so a single
            codepoint always fits in one fragment and the boundary back-off
            always makes progress).

    Returns:
        list[str]: The fragments. Empty list for empty input.

    Raises:
        ValueError: If `size` is too small to guarantee progress.
    """
    if not isinstance(size, int) or size < 4:
        # A UTF-8 codepoint is at most 4 bytes, so size >= 4 guarantees the
        # boundary back-off below always makes progress.
        raise ValueError("fragment size must be an integer >= 4")

    data = text.encode("utf-8")
    total = len(data)
    chunks: list = []
    start = 0

    while start < total:
        end = min(start + size, total)
        if end < total:
            # Back off over continuation bytes so the cut lands between
            # codepoints.
            while end > start and (data[end] & _UTF8_CONTINUATION_MASK) == _UTF8_CONTINUATION_PATTERN:
                end -= 1
        # Strict decode: safe by construction, and loud if ever violated.
        chunks.append(data[start:end].decode("utf-8"))
        start = end

    return chunks


def max_total_fragments(fragment_size: int, local_limit_bytes=None) -> int:
    """
    Upper bound on the number of fragments a single inbound stream may
    declare before it is rejected.

    When a positive local clipboard size limit is configured, the bound is
    derived from it (so legitimate transfers under the limit always fit);
    otherwise the hard ceiling applies.
    """
    if (
        isinstance(local_limit_bytes, (int, float))
        and not isinstance(local_limit_bytes, bool)
        and local_limit_bytes > 0
    ):
        derived = math.ceil(local_limit_bytes / fragment_size) + MAX_TOTAL_FRAGMENTS_MARGIN
        return max(MAX_TOTAL_FRAGMENTS_FLOOR, min(derived, MAX_TOTAL_FRAGMENTS_HARD))
    return MAX_TOTAL_FRAGMENTS_HARD


def is_valid_fragment_metadata(total_fragments, fragment_index) -> bool:
    """
    True when total_fragments / fragment_index are well-formed integers
    describing a position inside the stream. Rejects JSON floats, bools
    (which are ints in Python), and out-of-range indices.
    """
    if type(total_fragments) is not int or type(fragment_index) is not int:
        return False
    if total_fragments < 1:
        return False
    return 0 <= fragment_index < total_fragments
